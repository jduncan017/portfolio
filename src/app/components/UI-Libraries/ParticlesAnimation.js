"use client";
import { useEffect, useMemo, useState } from "react";
import Particles, { initParticlesEngine } from "@tsparticles/react";
import { loadSlim } from "@tsparticles/slim";

export const ParticlesAnimation = () => {
  const [init, setInit] = useState(false);
  const [particleLoad, setParticleLoad] = useState(false);

  useEffect(() => {
    initParticlesEngine(async (engine) => {
      await loadSlim(engine);
    }).then(() => {
      setInit(true);
    });
  }, []);

  const particlesLoaded = (container) => {
    setParticleLoad(true);
  };

  const options = useMemo(
    () => ({
      fullScreen: { enable: false },
      background: {
        image: "url('/bg1.jpg')",
        position: "50% 50%",
        repeat: "no-repeat",
        size: "cover",
        opacity: 1,
      },
      backgroundMask: {
        composite: "destination-out",
        cover: {
          color: {
            value: "#00060d",
          },
          opacity: 1,
        },
        enable: true,
      },
      clear: true,

      fpsLimit: 60,
      interactivity: {
        events: {
          onHover: {
            enable: true,
            mode: "repulse",
          },
        },
        modes: {},
      },
      particles: {
        move: {
          direction: "none",
          enable: true,
          outModes: {
            default: "bounce",
          },
          random: false,
          speed: 1,
          straight: false,
        },
        number: {
          density: {
            enable: true,
            height: 600,
          },
          value: 150,
        },
        opacity: {
          value: 1,
        },
        shape: {
          type: "circle",
        },
        size: {
          value: {
            min: 1,
            max: 30,
          },
        },
      },
      detectRetina: true,
      pauseOnBlur: true,
      pauseOnOutsideViewport: true,
    }),
    [],
  );

  if (init) {
    return (
      <div
        className={`ParticlesContainer transition-all duration-[2s] ${
          particleLoad
            ? "absolute left-0 top-0 h-full w-full overflow-hidden opacity-100"
            : "opacity-0"
        }`}
      >
        <Particles
          className="Particles h-full"
          id="tsparticles"
          particlesLoaded={particlesLoaded}
          options={options}
        />
      </div>
    );
  }

  return <></>;
};
