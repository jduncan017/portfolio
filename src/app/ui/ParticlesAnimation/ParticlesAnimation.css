.particles-outer-container {
	height: 100%;
	left: 0;
	overflow: hidden;
	position: absolute;
	top: 0;
	width: 100%;
}

.particles {
	height: 100%;
}