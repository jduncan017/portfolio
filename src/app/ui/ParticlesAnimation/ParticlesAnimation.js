import React, { useEffect, useState } from 'react';
import Particles from 'react-particles-js';
import './ParticlesAnimation.css';

export const ParticlesAnimation = () => {

  const [ numParticles, setNumParticles ] = useState(120)

  useEffect(() => {
    window.addEventListener('resize', updateParticles)
    updateParticles()
  })

  const updateParticles = () => {
    const viewportWidth = window.innerWidth
    if (viewportWidth > 400) setNumParticles(120)
    else setNumParticles(200)
  }

  return (
    <section className='particles-outer-container'>

      <Particles
        className='particles'
        params={{
          particles: {
            number: {
              value: numParticles,
              density: {
                enable: true,
                value_area: 1000,
              },
            },
            size: {
              value: 3,
            },
            move: {
              speed: 0.8,
            },
            color: {
              value: '#752082',
            },
            opacity: {
              value: 0.9,
              anim: {
                enable: true,
              },
            },
            line_linked: {
              enable: true,
              distance: 150,
              color: '#F600FF',
              opacity: 0.4,
              width: 2,
            },
          },
          interactivity: {
            events: {
              onhover: {
                enable: true,
                mode: 'repulse',
              },
            },
          }
        }} 
      />

    </section>
  );
}