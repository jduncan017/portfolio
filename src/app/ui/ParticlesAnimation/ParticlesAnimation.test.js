import React from 'react';
import { ParticlesAnimation } from './ParticlesAnimation';
import { shallow } from 'enzyme';

describe('ParticlesAnimation', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ParticlesAnimation />);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

});
