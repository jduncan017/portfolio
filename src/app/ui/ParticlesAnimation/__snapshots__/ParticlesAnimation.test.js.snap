// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticlesAnimation should match the snapshot 1`] = `
<section
  className="particles-outer-container"
>
  <f
    className="particles"
    height="100%"
    id="tsparticles"
    params={
      Object {
        "interactivity": Object {
          "events": Object {
            "onhover": Object {
              "enable": true,
              "mode": "repulse",
            },
          },
        },
        "particles": Object {
          "color": Object {
            "value": "#752082",
          },
          "line_linked": Object {
            "color": "#F600FF",
            "distance": 150,
            "enable": true,
            "opacity": 0.4,
            "width": 2,
          },
          "move": Object {
            "speed": 0.8,
          },
          "number": Object {
            "density": Object {
              "enable": true,
              "value_area": 1000,
            },
            "value": 120,
          },
          "opacity": Object {
            "anim": Object {
              "enable": true,
            },
            "value": 0.9,
          },
          "size": Object {
            "value": 3,
          },
        },
      }
    }
    style={Object {}}
    width="100%"
  />
</section>
`;
